### PR TITLE
sandbox: Add a command line flag to disable DAML stack traces

### PR DESCRIPTION
**Pull request by hurryabit**
The sandbox now accepts a `--stack-traces no` flag which will turn off
the location tracking in DAML Engine required to produce stack traces
for failing DAML code.

Benchmarks suggest that DAML Engine spends about 10% of its time with
tracking locations. Thus, this flag will give us roughly a 1.1x
speedup when stack traces are not needed.

This flag is still hidden because we would like to validate its
usefulness before we commit to supporting it.

CHANGELOG_BEGIN
CHANGELOG_END
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

---
This change is [Reviewable](https://reviewable.io/reviews/digital-asset/daml/6512)

### CODE REVIEW

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ConcurrentCompiledPackages.scala [hunk 1]
```diff
@@ -23,16 +23,20 @@ final class ConcurrentCompiledPackages extends MutableCompiledPackages {
...
 
-  def stackTraceMode = speedy.Compiler.FullStackTrace
   def profilingMode = _profilingMode
   def profilingMode_=(profilingMode: speedy.Compiler.ProfilingMode) = {
     _profilingMode = profilingMode
   }
+  def stackTraceMode = _stackTraceMode
+  def stackTraceMode_=(stackTraceMode: speedy.Compiler.StackTraceMode) = {
+    _stackTraceMode = stackTraceMode
+  }
```
Comment (SamirTalwar-DA) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ConcurrentCompiledPackages.scala:39]: Why do we use explicit trivial getters and setters instead of a public `var`? Scala should generate them for us.

Comment (hurryabit) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ConcurrentCompiledPackages.scala:39]: `stackTraceMode` is a method from the `CompiledPackages` trait this class implements. It's very unfortunate that Scala's syntax does not make it explicit that I'm overriding a method here. 😦

Comment (SamirTalwar-DA) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ConcurrentCompiledPackages.scala:39]: Makes sense! You can add `override` as prefix to `def` to make this clear. (And if there isn't a good reason not to, please do.)

Comment (SamirTalwar-DA) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ConcurrentCompiledPackages.scala:39]: Oh, it looks like you can just do:

```
  override var stackTraceMode = speedy.Compiler.FullStackTrace
```

This will add the setter and override the getter.

Comment (hurryabit) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ConcurrentCompiledPackages.scala:39]: The latter fails to compile with
```
error: variable stackTraceMode overrides nothing
  override var stackTraceMode = _stackTraceMode
```
Changing the `def` to `override def` works though. I'll go with this.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala [hunk 2]
```diff
@@ -138,6 +138,7 @@ final class SandboxServer(
       Files.createDirectories(profileDir)
       engine.startProfiling(profileDir)
   }
+  engine.enableStackTraces(config.stackTraces)
```
Comment (hurryabit) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala:141]: @SamirTalwar-DA Is that enough to make it work in `sandbox-next`? Or does it only affect `sandbox-classic`?

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ConcurrentCompiledPackages.scala [hunk 3]
```diff
@@ -23,16 +23,20 @@ final class ConcurrentCompiledPackages extends MutableCompiledPackages {
...
   private[this] var _profilingMode: speedy.Compiler.ProfilingMode = speedy.Compiler.NoProfile
+  private[this] var _stackTraceMode: speedy.Compiler.StackTraceMode = speedy.Compiler.FullStackTrace
 
   def getPackage(pId: PackageId): Option[Package] = _packages.get(pId)
   def getDefinition(dref: speedy.SExpr.SDefinitionRef): Option[speedy.SExpr] =
     Option(_defns.get(dref))
 
-  def stackTraceMode = speedy.Compiler.FullStackTrace
-  def profilingMode = _profilingMode
+  override def profilingMode = _profilingMode
```
Comment (hurryabit) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ConcurrentCompiledPackages.scala:32]: This is unrelated. I've just learnt about from @SamirTalwar-DA in the change below and though it might be a good idea to _immediately_ apply my lesson here as well, given how small the PR is.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala (hunk 2 above)
Comment (SamirTalwar-DA) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala:141]: Oh, good catch. This will only affect Sandbox Classic. You'll need to add it to `com.daml.platform.sandboxnext.Runner` too.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/cli/Cli.scala [hunk 4]
```diff
@@ -303,6 +303,12 @@ object Cli {
         .action((dir, config) => config.copy(profileDir = Some(dir.toPath)))
         .text("Enable profiling and write the profiles into the given directory.")
 
+      opt[Boolean]("stack-traces")
+        .hidden()
+        .optional()
+        .action((enabled, config) => config.copy(stackTraces = enabled))
+        .text("Enable stack traces. Default is to enable them.")
```
Comment (nickchapman-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/cli/Cli.scala:310]: Is this text correct? It reads very odd.

Comment (nickchapman-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/cli/Cli.scala:310]: perhaps: `"Enable/Disable stack traces. Default is to enable them."`

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala (hunk 2 above)
Comment (hurryabit) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala:141]: Thanks for letting me know. I'll do that in a followup PR together with the profiler as well.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/cli/Cli.scala (hunk 4 above)
Comment (hurryabit) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/cli/Cli.scala:310]: Good idea. Thanks.

Comment (hurryabit) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/cli/Cli.scala:310]:
```suggestion
        .text("Enable/disable stack traces. Default is to enable them.")
```
